### DISCUSSION TITLE
Tiny Organization

### DISCUSSION THREAD

**Discussion by HSorensen** [GCC-Tiny/.github#1] [Ideas]
Hi @SamuraiCrow !

You basically read my mind when you created this org. Thank you for the invite, looking forward to this.

Let try to create a plan...

To bring Tiny up to current GCC versions how about something like this:

- [X] Fork current GCC repo
- [x] Bring existing Tiny blogs to TexInfo or similar formatting language
- [ ] Review and update existing Tiny blogs
- [ ] Figure out branch concept for each chapter in the articles
- [x] Figure out how to publish new blogs
- [ ] Plan additional blogs, like compiler options etc.

Thoughts?
-Henrik

**Comment by SamuraiCrow** [comment]
I've already configured the .github project to use Jekyll to export the MarkDown formatted documentation to HTML like a blog.  It's the same software the author used.  See https://gcc-tiny.github.io/.github/ as the current deployment.

**Reply by HSorensen** [reply]
great. I wonder if we can use https://import.jekyllrb.com/ for the blogs ?

**Reply by SamuraiCrow** [reply]
The original articles are in Jekyll format already.  The author's blog is also using Jekyll.

**Reply by HSorensen** [reply]
But do we have the source for the blogs?

**Reply by SamuraiCrow** [reply]
No.  I see you requested the source already though.

An addendum revisiting the original post with a link to that post would be trivially easy for now.

**Reply by SamuraiCrow** [reply]
[Pandoc](https://github.com/jgm/pandoc) can convert Markdown to TexInfo or any other common file format.

**Reply by SamuraiCrow** [reply]
https://github.com/rofirrim/gcc-tiny/issues/1#issuecomment-1585775400

Grants permission as long as original author gets credit for work.  Thanks to him!

**Reply by SamuraiCrow** [reply]
I've used pandoc to filter out most of the HTML formatting from the website but the syntax highlighting and line numbers were done  manually and after trying to approximate it using Pascal formatting, I'm not sure that is going to work.  I need a different way to apply style guides to the code.  Once done, applying a TexInfo conversion should be easier.

**Reply by HSorensen** [reply]
I created gcc-tiny-doc with a texi basic structure. To get started I copied the text directly from the websites and added to the individual files, surrounded with ignore / end ignore.

The file part00-main/part00.texi, includes all the other chapters. After porting the text to texi, we can start to think about where to publish.

**Reply by HSorensen** [reply]
We should probably use something like [CC BY-SA](https://creativecommons.org/licenses/by-sa/4.0/) ?

That seems to be inline with Roger's request.

**Reply by HSorensen** [reply]
As Copyrights:

Copyright © 2016 Roger Ferrer Ibáñez.
Copyright © 2023 GCC Tiny Project.

thoughts?

**Reply by SamuraiCrow** [reply]
Agreed!  The license and copyrights seem to fit.

If you check the branch list of the gcc-tiny-doc project, I've converted the webpages to Markdown in the 'pandoc' branch.  The first chapter was unsuccessfully hand converted the rest of the way.  The problem I ran into was the syntax highlighting being rendered manually to styles in the HTML format.  On some Markdown renderers, the HTML passthrough is considered acceptible.  I haven' tried using Pandoc to export to TexInfo yet.

I've tentatively pointed Jekyll at the results of the Markdown format.  See the hand converted Markdown at [https://gcc-tiny.github.io/gcc-tiny-doc/part1/](https://gcc-tiny.github.io/gcc-tiny-doc/part1/).  The headers from the original webpage can just be removed.

**Reply by HSorensen** [reply]
Just as a reference I ported part 1 into texinfo. Definitely doable, but still some fine tuning needed for better rendering results. See [texinfo part01](https://github.com/GCC-Tiny/gcc-tiny-doc/blob/firstpart/src/part01.texi).

**Reply by HSorensen** [reply]
So I think the pandoc version you have started will work as well. Let's start with that. The texinfo was more an experiment.

**Reply by HSorensen** [reply]
Another option to consider would be to use [read the docs](https://docs.readthedocs.io/en/stable/tutorial/).Seems like a good place for the content.

**Reply by HSorensen** [reply]
@SamuraiCrow are you ok with using https://github.com/readthedocs/tutorial-template/ ?
I really want to get this started, and gettingTiny blog and code up to the latest version of GCC. The blogs need some updates and it is easier to do together with the actual code.

**Reply by SamuraiCrow** [reply]
My only concern with changing the format to that of Sphinx is that the example Mini code needed syntax highlighting and line numbers.  The original blog used some funky techniques to overlay the syntax highlighted code on another column with line numbers.  Pandoc should do the conversion work once it's established that that works.

**Reply by HSorensen** [reply]
@SamuraiCrow I took a stab at it: https://gcc-tiny-docs-rtd.readthedocs.io/en/latest/
source in https://github.com/GCC-Tiny/gcc-tiny-docs-rtd
I think this will work well enough for this blog series.
If you agree with this please let me know your rtd username and I will add you as maintainer.
Thoughts?

**Reply by SamuraiCrow** [reply]
My new signup on RTD is SamuraiCrow, the same as my GitHub.  I'm eager to see how this turns out!

**Reply by HSorensen** [reply]
You might have to accept the invite as maintainer, but added you: please verify https://readthedocs.org/projects/gcc-tiny-docs-rtd/

**Reply by SamuraiCrow** [reply]
I'm in.  Invitation accepted.

**Reply by HSorensen** [reply]
Great. 
Looking at the results, I think we can use RTD.

I did not do anything wrt build automation, license and copyrights etc.

I will start the part02 and let's see how it goes.

**Reply by HSorensen** [reply]
Added first stab at [part02.rst](https://github.com/GCC-Tiny/gcc-tiny-docs-rtd/blob/parts/docs/source/part02.rst),  [part03.rst](https://github.com/GCC-Tiny/gcc-tiny-docs-rtd/blob/parts/docs/source/part03.rst) and [part04.rst](https://github.com/GCC-Tiny/gcc-tiny-docs-rtd/blob/parts/docs/source/part04.rst)
The [results](https://gcc-tiny-docs-rtd.readthedocs.io/en/latest/) still need some fine tuning: 
- Sphinx' parts vs chapter vs sections, make sure the TOC is reflecting how the docs are structured
- Proper cross inline references for educational purposes
- make the writing style consistent with the change from a blog into a document
- adding license file CC-BY-SA

Once we have blog imported into RTD, I will start to do the updates to bring Tiny to latest version of GCC.

**Reply by SamuraiCrow** [reply]
You're making good progress!  Do you need me to help with anything?  I probably should be doing careful proofreading soon.

**Reply by HSorensen** [reply]
I will continue with the first rough conversion. There are some pictures and links we need to add as well as proof reading.
One thing. if you could look at the RTD ci/cd that would be helpful. I just took the tutorial template but did not look at any automation.

**Reply by SamuraiCrow** [reply]
The WebHook was already installed on the template.  It looks like it's already running the automations.

**Reply by HSorensen** [reply]
heh, so they are. Great.

**Reply by HSorensen** [reply]
@SamuraiCrow so I finished all the parts. Still need a review for consistency across all the different parts.

**Reply by SamuraiCrow** [reply]
Ok.  I've been gone a few days last week but this week looks good.  I'll give things a look through.

**Comment by SamuraiCrow** [comment]
OK.  I think I'll throw out the changes I made to the syntax highlights in the section 1 Markdown.  It's probably easier just to implement the style sheet.

**Comment by HSorensen** [comment]
@SamuraiCrow how did you create the build workflow? Is there an edit or view of the actions?

**Reply by SamuraiCrow** [reply]
I downloaded the HTML source of each page using the Brave browser and ran pandoc on each page using the HTML5 source document type and the pandoc version of Markdown as the destination.  Pandoc itself was downloaded from the Manjaro Linux package manager.

**Reply by SamuraiCrow** [reply]
## Update
I have reverted the changes to part1.  The preview is to look at the result of the GitHub Pages hosting.  See [here](https://gcc-tiny.github.io/gcc-tiny-doc/) for the rendered webpages.  Just stripping the header and footer should do it for a more finalized draft (apart from the syntax highlighting I mentioned earlier).

**Comment by HSorensen** [comment]
@SamuraiCrow I think we can close this discussion. After a few tries we landed on a got place using RTD. For additional topic let have a separate discussion where we can share research and ideas.

**Reply by SamuraiCrow** [reply]
Agreed.
